tern_api/reports.py

```diff
@@ -3,3 +3,194 @@
 #
 # Copyright (c) 2022 VMware, Inc. All Rights Reserved.
 # SPDX-License-Identifier: BSD-2-Clause
+
+import json
+import logging
+import os
+import subprocess
+import sys
+from dataclasses import asdict, dataclass, field
+from typing import Any, Dict, Optional
+from uuid import uuid4
+
+from tern_api import tern_app, tern_tasks
+from tern_api.constants import task_status
+from tern_api.utils import TernAPIResponse
+
+
+class TernError(Exception):
+    """Failure on the Tern execution."""
+
+
+@dataclass
+class DataResponse:
+    id: str
+    cache: bool = field(default=True)
+    message: str = field(default="")
+    status: str = field(default=task_status.PENDING.value)
+    report: Dict[str, Any] = field(default_factory=dict)
+
+    def to_dict(self):
+        """Returns the DataResponse as a dictionary."""
+        return asdict(self)
+
+
+def tern(command: list) -> Dict[str, Any]:
+    """
+    Runs the tern CLI using the ``subprocess``.
+
+    Args:
+        command: Command line as a list (required by subprocess).
+
+    Returns:
+       report as a dictionary (JSON).
+
+    Raises:
+        TernError: failure during running the command from tern CLI.
+    """
+    logging.debug(command)
+
+    if type(command) != list:
+        raise TypeError("command must be a list")
+
+    tern_cmd = subprocess.run(command, capture_output=True)
+    if tern_cmd.stdout:
+        json_report = json.loads(tern_cmd.stdout)
+        return json_report
+    else:
+        if tern_cmd.stderr:
+            logging.debug(tern_cmd.stderr)
+
+            # Is important to return a specific error if the error comes from
+            # tern CLI, it means the task FINISH, for example, invalid image.
+            logging.info(tern_cmd.stderr)
+            raise TernError(tern_cmd.stderr.decode())
+
+
+@tern_tasks.job
+def tern_report(
+    command: list, cache: bool, cache_file: Optional[str]
+) -> Dict[str, Any]:
+    """
+    Tern Report as a task (background) and manages the cache
+
+    Args:
+        command: Command line as a list (required by subprocess).
+        cache: Use caching
+        cache_file: If cache, inform the cache_file
+
+    Return:
+        Report as Dictionary
+    """
+    if cache:
+        # If a API user is using the cache, first try to load the cached
+        # instead doing a new call to the tern.
+        try:
+            with open(cache_file, "r") as f:
+                report = json.load(f)
+            return report
+        except FileNotFoundError:
+            logging.debug(f"Cache file not found: {cache_file}")
+            # call the tern and dump it to the cache
+            report = tern(command)
+            with open(cache_file, "w") as f:
+                json.dump(report, f, indent=2)
+    else:
+        report = tern(command)
+
+    return report
```

Comment:
You can modify this function logic.
Instead of calling ``tern(command)``, it could call the tern library API. 
Also, modify the ``submit()`` to build the needed arguments for the tern library API instead ``command``.

Having It in a specific function (see ``tern`` line 38) is even better to handle the possible exceptions cases and make task management consistent.

Reply:
Any suggestions on how to test the API other than using curl from the command line?